benchmarks/lasso/solvers/baseline.py

```diff
@@ -5,33 +5,34 @@
 
 
 class Solver(BaseSolver):
-    name = 'Baseline'
+    name = 'Baseline'  # proximal gradient, eventually accelerated
 
+    # any parameter defined here is accessible as a class attribute
     parameters = {'use_acceleration': [False, True]}
 
     def set_objective(self, X, y, lmbd):
         self.X, self.y, self.lmbd = X, y, lmbd
 
-        self.L = np.linalg.norm(self.X.dot(self.X.T), ord=2)
+        self.L = np.linalg.norm(self.X, ord=2) ** 2
 
     def run(self, n_iter):
         n_features = self.X.shape[1]
-        z_hat = np.zeros(n_features)
+        w = np.zeros(n_features)
         t_new = 1
         for _ in range(n_iter):
             if self.use_acceleration:
                 t_old = t_new
                 t_new = (1 + np.sqrt(1 + 4 * t_old ** 2)) / 2
-                z_old = z_hat.copy()
-            grad = self.X.T.dot(self.X.dot(z_hat) - self.y)
-            z_hat -= 1 / self.L * grad
-            z_hat = self.st(z_hat, 1 / self.L * self.lmbd)
+                w_old = w.copy()
+            grad = self.X.T.dot(self.X.dot(w) - self.y)
+            w -= 1 / self.L * grad
+            w = self.st(w, 1 / self.L * self.lmbd)
             if self.use_acceleration:
-                z_hat = z_hat + (t_old - 1.) / t_new * (z_hat - z_old)
-        self.z_hat = z_hat
+                w = w + (t_old - 1.) / t_new * (w - w_old)
+        self.w = w
 
-    def st(self, z, mu):
-        return np.sign(z) * np.maximum(0, np.abs(z) - mu)
+    def st(self, w, mu):
+        return np.sign(w) * np.maximum(0, np.abs(w) - mu)
```

Comment:
st could operate inplace to avoid unecessary memory allocations

Reply:
I thought of using `shrink = np.maximum(0, 1 - mu / np.abs(w))` but this causes division by 0 and working with masks causes creation of other arrays.

In hindsight the gradient should be the larger cost of the iteration, have you observed the contrary in the past?

Reply:
i doub't know how much impact this will have true. But here you make 4 copies of w in one line

Reply:
```In [5]: a = np.array([0., 1.])                                                  

In [6]: a *= np.maximum(0, 1 - 0.5 / np.abs(a))                                 
/home/mathurin/miniconda3/bin/ipython:1: RuntimeWarning: divide by zero encountered in true_divide
  #!/home/mathurin/miniconda3/bin/python

In [7]: a                                                                       
Out[7]: array([0. , 0.5])
```
but this raises a warning bc of mu / 0

I think hopefully numpy reuses temp arrays now and only 2 arrays are needed

Reply:
Inplace soft thresholding:
```python
def st(w, mu):
    a -= np.clip(w, -mu, mu)
    return a
```